sdk/Sdk.Generators/FunctionMetadataProviderGenerator.Emitter.cs

```diff
@@ -0,0 +1,167 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.CodeDom.Compiler;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.Azure.Functions.Worker.Sdk.Generators
+{
+    public partial class FunctionMetadataProviderGenerator
+    {
+        internal sealed class Emitter
+        {
+
+            public string Emit(IReadOnlyList<GeneratorFunctionMetadata> funcMetadata, CancellationToken cancellationToken)
+            {
+                using var stringWriter = new StringWriter();
+                using (var indentedTextWriter = new IndentedTextWriter(stringWriter))
+                {
+                    var hasHttpTrigger = funcMetadata.Any(funcMetadata => funcMetadata.IsHttpTrigger);
+
+                    SetUpUsings(indentedTextWriter, hasHttpTrigger);
+
+                    // create namespace
+                    indentedTextWriter.WriteLine("namespace Microsoft.Azure.Functions.Worker");
+                    indentedTextWriter.WriteLine("{");
+                    indentedTextWriter.Indent++;
+
+                    // create class that implements IFunctionMetadataProvider
+                    indentedTextWriter.WriteLine("public class GeneratedFunctionMetadataProvider : IFunctionMetadataProvider");
+                    indentedTextWriter.WriteLine("{");
+                    indentedTextWriter.Indent++;
+
+                    WriteGetFunctionsMetadataAsyncMethod(indentedTextWriter, funcMetadata, cancellationToken);
+
+                    indentedTextWriter.Indent--;
+                    indentedTextWriter.WriteLine("}");
+
+                    // add method that users can call in startup to register the source-generated file
+                    AddRegistrationExtension(indentedTextWriter);
+
+                    indentedTextWriter.Indent--;
+                    indentedTextWriter.WriteLine("}");
+
+                    indentedTextWriter.Flush();
+                }
+
+                return stringWriter.ToString();
+            }
+
+            private void SetUpUsings(IndentedTextWriter indentedTextWriter, bool hasHttpTrigger)
+            {
+                indentedTextWriter.WriteLine("// <auto-generated/>");
+                indentedTextWriter.WriteLine("using System;");
+                indentedTextWriter.WriteLine("using System.Collections.Generic;");
```

Comment:
Ignore this comment. Because one source gen should not rely on results from another source gen as there is no guarantee on the order of execution(of source generators).
Keep this code.